File: java-samples/src/main/java/com/example/sample/member/MemberService.java
package com.example.sample.member;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * [Week 2] Service 변환 과제
 *
 * 변환 포인트:
 * - @Service → @Service (kotlin-spring allopen 없으면 @Transactional 프록시 생성 실패)
 * - @Transactional 클래스/메서드 레벨 선택
 * - Optional.orElseThrow() → ?: throw 로 교체 가능
 * - 중복 체크 패턴: if + throw → require / check
 */
@Service
@Transactional(readOnly = true)
public class MemberService {

    private final MemberRepository memberRepository;

    public MemberService(MemberRepository memberRepository) {
        this.memberRepository = memberRepository;
    }

    @Transactional
    public SignupResult signup(MemberSignupRequest request) {
        if (memberRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("이미 사용 중인 이메일입니다: " + request.getEmail());
        }

        Member member = Member.signup(
                request.getEmail(),
                request.getName(),
                request.getPassword()
        );

        Member saved = memberRepository.save(member);
        return SignupResult.success(saved.getEmail());
    }

    public MemberResponse getById(Long id) {
        Member member = memberRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("회원을 찾을 수 없습니다. id=" + id));

        return new MemberResponse(id, member.getEmail(), member.getName());
    }
}

File: kotlin-lab/src/main/kotlin/com/example/lab/domain/member/MemberService.kt
package com.example.lab.domain.member

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class MemberService(
    private val memberRepository: MemberRepository,
    private val memberProperties: MemberProperties
) {

    lateinit var lastSignupEmail: String

    val configSummary: String by lazy {
        println("configSummary 계산 중")
        "default=${memberProperties.defaultStatus}"
    }

    @Transactional
    fun signup(memberSignupRequest: MemberSignupRequest): SignupResult {
        println("memberProperties : ${memberProperties}")

        try {
            println("lastSignupEmail (init 전) : ${lastSignupEmail}")
        } catch (e: UninitializedPropertyAccessException) {
            println("lateinit 예외 : ${e.message}")
        }

        lastSignupEmail = memberSignupRequest.email
        println("lastSignupEmail (init 후) : ${lastSignupEmail}")

        println("configSummary 1차 호출 : ${configSummary}")
        println("configSummary 2차 호출 : ${configSummary}")

        if (memberRepository.existsByEmail(memberSignupRequest.email)) {
            throw IllegalArgumentException("이미 사용 중인 이메일입니다:  ${memberSignupRequest.email}")
        }
        val member: Member = Member.signup(
            memberSignupRequest.email,
            memberSignupRequest.name,
            memberSignupRequest.password
        )

        memberRepository.save(member)
        return SignupResult.Success

    }

    @Transactional(readOnly = true)
    fun getById(id: Long): MemberResponse {
        val member = memberRepository.getById(id)
        // 저장된 id는 null 일 수 없으니까 !!
        return MemberResponse(member.id!!, member.email, member.name)
    }
}